### DISCUSSION TITLE
"@as_declarative()" combined with DBSession for multiple (existing) dbs

### DISCUSSION THREAD

**Discussion by UnravelSports** [sqlalchemy/sqlalchemy#7556] [Usage Questions] [answered]
Hi,

I'm looking to set up multiple database support for FastAPI using SQLAlchemy for two existing databases. Let me start of by saying that I've tried a whole bunch of different approaches, but I can't seem to get it to work. I'm trying to use a replacement for `Base = declarative_base()` since it only supports a single db, with some form of `@as_declarative()`. I've been trying to implement the following example/answer I found in the FastAPI discussion board https://github.com/tiangolo/fastapi/issues/2592#issuecomment-773208039, but it feels like this might be more of a SQLAlchemy question (considering I've been reading mostly those docs to try to resolve my issue). 

Below is a short version of the code I'm tyring to implement:

    from config import *
    
    from sqlalchemy.ext.declarative import as_declarative
    from sqlalchemy import create_engine, Table, MetaData, event
    
    
    host = os.environ.get("DB_HOST")
    port = os.environ.get("DB_PORT")
    
    user_1 = os.environ.get("DB_USER_1")
    password_1 = os.environ.get("DB_PASSWORD_1")
    
    password_2 = os.environ.get("DB_PASSWORD_2")
    user_2 = os.environ.get("DB_USER_2")
    
    SQLALCHEMY_DATABASE_URL_1 = f"postgresql://{user_1}:{password_1}@{host}:{port}/postgres"
    SQLALCHEMY_DATABASE_URL_2 = f"postgresql://{user_2}:{password_2}@{host}:{port}/postgres"
    engine_1 = create_engine(SQLALCHEMY_DATABASE_URL_1)
    engine_2 = create_engine(SQLALCHEMY_DATABASE_URL_2)
    
    def init_search_path(connection, conn_record):
        cursor = connection.cursor()
        try:
            cursor.execute('SET search_path TO table_1;')
        finally:
            cursor.close()
    
    event.listen(engine_1, 'connect', init_search_path)
    
    @as_declarative()
    class ModelDB1(object):
        __tablename__ = 'users'
        __table__ = Table(
           __tablename__,
            MetaData(),
            autoload=True,
            autoload_with=engine_1
        )
    
    @as_declarative()
    class ModelDB2(object):
        __tablename__ = 'schedules',
        __table__ = Table(
            __tablename__,
            MetaData(),
            autoload=True,
            autoload_with=engine_2
        )
    
    DBSession = sessionmaker(
        autocommit=False, autoflush=False
    )
    DBSession.configure(binds={
        ModelDB1: engine_1,
        ModelDB2: engine_2
    })
    
    def get_db():
        db = DBSession()
        try:
            yield db
        except:
            db.rollback()
        finally:
            db.close()
    
    @app.get("/1/users", status_code=200)
    async def get_users(db: Session = Depends(get_db)):
        query = f"""
               SELECT SELECT * FROM table_1.users     
        """
        execution = db.execute(text(query))
        # do things
        return {"1": "success"}
    
    #/team?opta_team_id=6238
    @app.get("/2/schedules", status_code=200)
    async def get_schedules(db: Session = Depends(get_db)):
        query = f"""
               SELECT * FROM table_2.schedules
        """
        execution = db.execute(text(query))
        # do things
        return {"2": "success"}

In the answer linked above, I had a feeling something as missing in the `ModelDB` classes, so I did some digging and found [this stackoverflow answer](https://stackoverflow.com/questions/39955521/sqlalchemy-existing-database-query) because without declaring anything in the classes I was getting an `UnboundExecutionError` error. Since my tables already exist I wanted to find an implementation that could help me with that (hence that SO answer).

Unfortunately using both `__tablename__ = table_1.users`, `__tablename__ = table_1` and `__tablename__ = users` (and same for `table_2`) I got an error stating that the tables do not exist.

So, I tried to use the answer to [this SO question](https://stackoverflow.com/questions/18076515/sqlalchemy-nosuchtableerror-when-table-exists-in-database), to help me alleviate that problem, but neither of the two approaches mentioned in the selected answer do anything for me.

Does anyone have any idea where I'm going wrong here?

**Comment by CaselIT** [comment, marked as answer]
Hi,

I think the issue here is that SQLAlchemy does not process the `text` you are passing to execute the queries, so it does not know which engine to use.
You can either use a non textual select or create two sessions each bound to a different engine and use one or the other.

We could probably improve the documentation mentioning that `text` is not supported

**Reply by CaselIT** [reply]
You may also use [`bind_arguments`](https://docs.sqlalchemy.org/en/14/orm/session_api.html?highlight=session#sqlalchemy.orm.Session.execute.params.bind_arguments) to tell the session which engine to use

**Reply by UnravelSports** [reply]
Thank you for the reply!

I saw `bind_arguments` mentioned a couple of times in the docs, but I couldn't find anywhere how to use them, or where to pass them or what it should look like. Would you mind explaining?

And about your first answer, you mean that `db.execute(text(query))` should simply be `db.execute(query)?`

**Reply by CaselIT** [reply]
> I saw `bind_arguments` mentioned a couple of times in the docs, but I couldn't find anywhere how to use them, or where to pass them or what it should look like. Would you mind explaining?

Since sqlalchemy does not parse the text query, it cannot know which engine to use.  `bind_arguments` can be used to specify a mapper that's used in the text columns, so that it can use that to select the engine. Using `db.execute(text(query), bind_arguments={'mapper': ModelDB2})` will tell sqlalchemy to use `ModelDB2` to select the engine



> And about your first answer, you mean that `db.execute(text(query))` should simply be `db.execute(query)?`

No, By "non textual select" I meant using sqlalchemy core functionality, so your query would be something like `sa.select('*').select_from(ModelDB2)`

**Reply by UnravelSports** [reply]
Thanks so much, now it all makes sense!
